homeassistant/components/fibaro/__init__.py

```diff
@@ -357,39 +355,75 @@ def _read_devices(self):
                 pass
 
 
-def setup(hass: HomeAssistant, base_config: ConfigType) -> bool:
-    """Set up the Fibaro Component."""
+async def async_setup(hass: HomeAssistant, base_config: ConfigType) -> bool:
+    """Migrate configuration from configuration.yaml."""
+    if DOMAIN not in base_config:
+        return True
     gateways = base_config[DOMAIN][CONF_GATEWAYS]
-    hass.data[FIBARO_CONTROLLERS] = {}
+    if gateways is None:
+        return True
+
+    # check if already configured
+    if hass.config_entries.async_entries(DOMAIN):
+        return True
 
-    def stop_fibaro(event):
-        """Stop Fibaro Thread."""
-        _LOGGER.info("Shutting down Fibaro connection")
-        for controller in hass.data[FIBARO_CONTROLLERS].values():
-            controller.disable_state_handler()
+    for gateway in gateways:
+        # prepare new config based on configuration.yaml
+        conf = {
+            CONF_URL: gateway[CONF_URL],
+            CONF_USERNAME: gateway[CONF_USERNAME],
+            CONF_PASSWORD: gateway[CONF_PASSWORD],
+            CONF_IMPORT_PLUGINS: gateway[CONF_PLUGINS],
+        }
+
+        # import into config flow based configuration
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+            )
+        )
 
-    hass.data[FIBARO_DEVICES] = {}
+    return True
+
+
+def _init_controller(data: dict[str, Any]) -> tuple[bool, FibaroController]:
+    """Validate the user input allows us to connect to fibaro."""
+    controller = FibaroController(data)
+    connected = controller.connect()
+    return connected, controller
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the Fibaro Component."""
+    connected, controller = await hass.async_add_executor_job(
+        _init_controller, entry.data
+    )
+    if not connected:
+        return False
```

Comment:
```suggestion
        raise ConfigEntryNotReady(f"Could not connect to controller at {entry.data[CONF_URL]}")
```
It would be better to still return False if its an auth error though, and in the future raise ConfigEntryAuthFailed to start a re-auth flow.